Mathlib/CategoryTheory/Galois/IsFundamentalgroup.lean

```diff
@@ -0,0 +1,284 @@
+      apply IsClosed.leftCoset
+      exact Subgroup.isClosed_of_isOpen _ (stabilizer_isOpen G X.pt)
+    · intro s
+      rw [Set.univ_inter]
+      obtain ⟨gs, hgs⟩ :=
+        toAut_surjective_isGalois_finite_family F G t (fun X : s ↦ X.val.obj) h
+      use gs
+      simp only [Set.mem_iInter]
+      intro X hXmem
+      rw [mem_leftCoset_iff, SetLike.mem_coe, MulAction.mem_stabilizer_iff, mul_smul,
+        hgs ⟨X, hXmem⟩, ← hgi X, inv_smul_smul]
+  obtain ⟨g, hg⟩ := hne
+  refine ⟨g, Iso.ext <| natTrans_ext_of_isGalois _ <| fun X _ ↦ ?_⟩
+  ext x
+  simp only [toAut_hom_app_apply]
+  have : g ∈ (gi ⟨X, x, inferInstance⟩ • MulAction.stabilizer G x : Set G) := by
+    simp only [Set.mem_iInter, c] at hg
+    exact hg _
+  obtain ⟨s, (hsmem : s • x = x), (rfl : gi ⟨X, x, inferInstance⟩ • s = _)⟩ := this
+  rw [smul_eq_mul, mul_smul, hsmem]
+  exact hgi ⟨X, x, inferInstance⟩ x
+
+/-- If `toAut F G` is surjective, then `G` acts transitively on the fibers of connected objects.
+For a converse see `toAut_surjective`. -/
+lemma isPretransitive_of_surjective (h : Function.Surjective (toAut F G)) (X : C)
+    [IsConnected X] : MulAction.IsPretransitive G (F.obj X) where
+  exists_smul_eq x y := by
+    obtain ⟨t, ht⟩ := MulAction.exists_smul_eq (Aut F) x y
+    obtain ⟨g, rfl⟩ := h t
+    exact ⟨g, ht⟩
+
+end
+
+section
+
+variable [GaloisCategory C]
+variable (G : Type*) [Group G] [∀ (X : C), MulAction G (F.obj X)]
+
+/-- A compact, topological group `G` with a natural action on `F.obj X` for each `X : C`
+is a fundamental group of `F`, if `G` acts transitively on the fibers of Galois objects,
+the action on `F.obj X` is continuous for all `X : C` and the only trivally acting element of `G`
+is the identity. -/
+class IsFundamentalGroup [TopologicalSpace G] [TopologicalGroup G] [CompactSpace G]
+    extends IsNaturalSMul F G : Prop where
+  transitive_of_isGalois (X : C) [IsGalois X] : MulAction.IsPretransitive G (F.obj X)
+  continuous_smul (X : C) : ContinuousSMul G (F.obj X)
+  non_trivial' (g : G) : (∀ (X : C) (x : F.obj X), g • x = x) → g = 1
+
+namespace IsFundamentalGroup
+
+attribute [instance] continuous_smul transitive_of_isGalois
+
+variable {G} [TopologicalSpace G] [TopologicalGroup G] [CompactSpace G] [IsFundamentalGroup F G]
+
+lemma non_trivial (g : G) (h : ∀ (X : C) (x : F.obj X), g • x = x) : g = 1 :=
+  IsFundamentalGroup.non_trivial' g h
+
+end IsFundamentalGroup
+
+variable [FiberFunctor F]
+
+/-- `Aut F` is a fundamental group for `F`. -/
+instance : IsFundamentalGroup F (Aut F) where
+  naturality g X Y f x := (FunctorToFintypeCat.naturality F F g.hom f x).symm
+  transitive_of_isGalois X := FiberFunctor.isPretransitive_of_isConnected F X
+  continuous_smul X := continuousSMul_aut_fiber F X
+  non_trivial' g h := by
+    ext X x
+    exact h X x
+
+variable [TopologicalSpace G] [TopologicalGroup G] [CompactSpace G] [IsFundamentalGroup F G]
+
+lemma toAut_bijective : Function.Bijective (toAut F G) where
+  left := by
```

Comment:
Could you extract a lemma `toAut_injective_of_non_trivial` similarly as for the surjectivity?